### PR TITLE
CustomTexture: new name format

### PR DESCRIPTION
**Pull request by degasus**
Warning: This will break the hashes for most paletted textures.

### CODE REVIEW

**Comment by degasus** [comment]
@BigheadSMZ have fun testing :D

**Comment by BigheadSMZ** [comment]
Just got home and this is taking priority over both shower and supper. I'll report back soon, so excited!!

**Comment by mimimi085181** [comment]
@BigheadSMZ :
If a palette uses the full available range, then the hashes from the old and the new method are supposed to be identical.
And also, there's a 3rd kind of paletted textures. This would either be 10 or A, depending on the number system being decimal or hexadecimal for this, which i don't know.

**Comment by degasus** [comment]
@BigheadSMZ Please don't use edits to update your post after a longer time. If there was a progress, just repost so I'll notice ;)

btw, the _10 format is still broken a lot. The code doesn't calculate the tlut size correct in this case :/

File: Source/Core/VideoCommon/HiresTextures.cpp [hunk 1]
```diff
@@ -82,6 +82,42 @@ void HiresTexture::Init(const std::string& gameCode)
...
+			{
+				min = std::min<u32>(min, texture[i]);
+				max = std::max<u32>(max, texture[i]);
+			}
+			break;
+		case 16384 * 2:
+			for (size_t i = 0; i < texture_size/2; i++)
+			{
+				min = std::min<u32>(min, ((u16*)texture)[i] & 0x3fff);
+				max = std::max<u32>(max, ((u16*)texture)[i] & 0x3fff);
```
Comment (degasus) [inline Source/Core/VideoCommon/HiresTextures.cpp]: TODO: Do we have to swap the byte order?

**Comment by degasus** [comment]
@BigheadSMZ Updated with lots of changes. Now all formats did change, but it automatically convert them on the fly. So please make a copy of your texture packs. Does everything still works fine?

**Comment by BigheadSMZ** [comment]
I don't know how the coding side of this looks, but from a users standpoint this is beyond my imagining. No more country code, the original dimensions in the texture, and the hashes seems longer making less chance for two textures to share a name. And the converter, wow. You really outdid yourself. I never imagined any of this when hoping to just see an end to duplicate textures. The original dimensions in the texture is so amazing, something so small but so incredibly convenient!

I have tried to replicate any bugs with the old method of dumping and have not been able to. No matter what options I tick with any accuracy in any order, a texture name is always dumped with the same name. Before I could get a total of 6 different possibilities per texture! Words can't properly express my gratitude for all of this, it feels almost too good and perfect to be true. I have not tested other texture packs yet, but for Xenoblade it's perfect.

The only thing I can think of that remains is allowing more levels of mipmaps, unless that is already done. And a question regarding mipmaps, is how small should the smallest mipmap be, like 8x8, or 4x4, or even all the way down to 1x1? I think once objects are that small the visual difference between 4 and 8 is almost imperceptible, but I should know what Dolphin expects for a minimum.

**Comment by BigheadSMZ** [comment]
Ok two things, one is issue with converter and other is something i remembered. 

I confirmed what StripTheSoul said here: https://forums.dolphin-emu.org/Thread-hd-retexture-xenoblade-chronicles-hd-texture-pack-v5-7-december-27-2014?pid=353730#pid353730 ... and it seems the converter misses the textures for Sonic Colors (SNCE8P). I redumped a few with Master branch Dolphin and they are indeed dumped correctly ("Fast" accuracy with "Load" checked). This is a very small pack, so redumping this wouldn't be much issue.. but the fact it is missing the textures for this game might mean it affects others.

Another thing I noticed Dolphin has always done is dump frames from movie sequences as textures. This happens in Xenoblade and Sonic Colors and I imagine any game with FMVs. It's not a huge issue if it has no work around, but it can be annoying because some videos in Xenoblade dump like 500 (maybe more?) frames. http://i.imgur.com/X2HI2gl.jpg

**Comment by degasus** [comment]
Using more levels per texture isn't done right now, but it won't affect this PR. So it should be fine to just create all levels right noew and it will be used soon. I'd suggest to go down to 1x1. It's not for a visual difference, it's to remove aliasing.
And damn, I've missed to also rename the mipmap textures... Might this be the issue with Sonic Colors? I'll look at this later :/

I'm sorry for videos. Technically, they are just textures. So I doubt a hack is worth to detect this kind of textures. Maybe you want to change the video? Just create a custom texture per frame :D

I'll add a flag if a texture has mipmaps. Sounds good.

**Comment by BigheadSMZ** [comment]
I fell asleep way too early last night... 
- For mipmaps, awesome! A flag would be really nice (maybe like tex_m_ or something simple, the names are already kind of long), and it's good to know what to use as smallest dimension, 1x1 is fine if it helps visually.
- Sonic Colors I'm not sure what's going on, I can't get a result I understand. It just ignores all 22 custom textures for this game for some reason and redumps them. The formats are _5 and _14 (and one _2), but these are working for Xenoblade.
- Don't worry about the videos, I was just bringing it up in case it was a Dolphin bug. But if they are textures and not an actual video format, then it makes sense that they dump.

So the only outstanding issues I can see is the mystery of the converter ignoring Sonic Colors, and whatever is left with mipmaps.

**Comment by degasus** [comment]
@BigheadSMZ You can check if the Sonic issue is the same as the mipmap issue if the textures without any "_mip%d" parts are converted but the _mapX textures did stay as they are.

**Comment by BigheadSMZ** [comment]
I tested packs for: Kirby Air Ride, Legend of Zelda: Wind Waker, Metal Gear Solid: Twin Snakes, New Super Mario Bros. Wii, Paper Mario: TTYD, Resident Evil 3, Star Wars Rogue Leader: Rogue Squadron II, Super Mario Sunshine, Super Paper Mario, Super Smash Bros. Brawl, Tales of Symphonia, and Xenoblade Chronicles.

All these packs are working as intended and the textures get converted. Sonic Colors is just being stubborn for some reason. Most of the textures in that pack are not ones that use mipmaps so I don't know about this game... It's only 22 textures total so it's not a big issue as all other packs seem to convert just fine.

And one last thing I can think of, if mipmap textures have a flag, is it still worth dumping mipmaps (mip1, mip2, etc)... As long as we know it needs mips and know the min dimensions should be 1, there is not much need to have the dumped mips.

**Comment by Tinob** [comment]
some games use mipmaps to achive efects like distance atenuetion, so they can be diferent from what you would spect.

**Comment by degasus** [comment]
"_m_" is now in the filename for mipmap textures.
Converting of mipmaps is implemented but untested right now.

@BigheadSMZ I don't think only dumping the first mipmap is worth. Dumping isn't only for custom texture, it's also for debugging.

How shall we choose if textures should be converted? Is just an ini fine or do you want me to create a gui checkbox?

Still to be discussed: Which hash algorithm shall we use? At the moment we can switch for free.

Shall we also mark this texture in some way which let us detect which format it has? Maybe we're able to change it easier in the future.

**Comment by Tinob** [comment]
the hash algorithm is a dificult issue, hardware crc is really fast in intel cpu but gives more colitions than murmurhash3, mumurhash3 is faster than crc on anything older than hashwell and on every amd cpu, xxhash shares the same properties of murmur3 and is the fastest, at least on amd and pre-hashwell, for new intesl i really don't have the oportunity to compare.

**Comment by JMC47** [comment]
The new CRC is definitely faster on Ivy Bridge and Sandy Bridge, according to my benchmarking.

**Comment by degasus** [comment]
Mipmaps seems to work fine, also Sonic seems to work fine here, but I have no clue what I've fixed.

**Comment by degasus** [comment]
Updated with a gfx ini option: "ConvertHiresTextures"
Used "tex1_" as prefix to support new formats in future ("tex2_"...)

I personally have no clue about hashes, so just tell me which one should be used :D

Ready for review?

**Comment by BigheadSMZ** [comment]
Tested several texture packs now and everything seems to working. If you are still open to suggestions, I have only two at this point. 

- It seems all new hashes for all textures have eight 0's in front of them, which seems redundant and just makes the name 8 characters longer (16 for paletted). I don't know if you have something in mind for that or it was intentional but its fine the way it is if it needs to be that way.

- Second is just a small nitpick and is not important either, but when looking at the textures in a list view having the m after the dimensions for mip textures might be slightly more visually appealing because the numbers line up.

tex1_256x164_00000000cd3c5907_3
tex1_m_256x256_000000009340a099_14
>> vs <<
tex1_256x164_00000000cd3c5907_3
tex1_256x256_m_000000009340a099_14

Both of these suggestions are not important at all and I'm happy it works so well and would be content forever in its current state. Thank you for all your hard work, Dolphin and its team never fail to amaze me month after month...

**Comment by degasus** [comment]
Oh damn, this zeros are a major issue. On linux, they're dumped with 16 digits - but easy to fix: PRIx64...

**Comment by degasus** [comment]
Both issues should be fixed, build is updated. Are the hashes now 16 digits long?

**Comment by BigheadSMZ** [comment]
Yes this format seems perfect now, except maybe typo for underscore in mipmap?

most: tex1_36x36_4653193aaa14d048_0.png
paletted: tex1_200x40_63f3408997969cd6_9a81ceb434acb873_8.png
mipmap: tex1_256x256m__efef240984783a47_14.png

Honestly it looks better that way except the double underscore so maybe this a bonus. Aside from that, I got nothing left at this point, I don't think it can be made much better than this. :)

**Comment by degasus** [comment]
yeah, it was a typo. It's hard to test such changes with only a webbrowser and the github edit interface...

**Comment by BigheadSMZ** [comment]
Tested it out again with the mipmap marker fix, and everything is beautiful. A tear of happiness ran down my cheek. I can't find any more bugs and the converter is working flawlessly with every situation I throw at it. Once again, even if you get sick of hearing it... amazing and thank you!

I would like to make a post in the custom texture section of the forums after this gets merged to master to notify other texture pack authors, unless someone else has this planned. Because I don't understand everything at the level of everyone here, I would appreciate it if someone can verify everything is correct, I'm not missing anything, or if there are better words for what I'm trying to say. I have a rough draft here: http://pastebin.com/FcP3Rsf7

**Comment by ZephyrSurfer** [comment]
Are the names of the dumped mipmap textures files smaller?
And if so is that on purpose?

**Comment by MaJoRoesch** [comment]
@BigHeadSMZ: This will be in the progress report, for sure. People will know about it!

**Comment by JMC47** [comment]
When we make a great feature like this more accessible for people to work on, I think it's definitely worth mentioning.  I'd love to see more Texture Packs in the future.  Love the RS2/Xenoblade packs so far.

**Comment by degasus** [comment]
I've uploaded your proposel to google drive to allow comments: https://docs.google.com/document/d/1NyjYv2vhLAzqDXoKreKVs32jNl1ka5Tr0VXUsm8PvVk/edit?usp=sharing

**Comment by degasus** [comment]
Maybe we should also create a small Howto for new texture packs. eg integer upscaling factor, create all minimaps if possible, ...

**Comment by Tinob** [comment]
Excelent work, is looking really great. About the hash giving it a second view we can't use crc because it depends on a specific architecture and we want support for hires textures on android to. So it reduced to compare murmur vs xxhash. If someone is interested in testing I can add xxhash to the basecode for benchmarking.

**Comment by degasus** [comment]
@Tinob that sounds fine. I guess xxhash would be the best.

**Comment by ZephyrSurfer** [comment]
@Tinob  I have a G3258 so I know some Haswell extensions aren't included but I would like to see that to make a comparison anyway.

**Comment by Tinob** [comment]
ok here you have
https://mega.co.nz/#!0ksE3ZzC!usUQ4yyaZ45VcB6GzvKu5Vc0wPb4boXcfppaC6kuLO0
this version add xxhash and a configuration option to set the desired hash function.
all hash function have been forced to do a complete hash so sample slider will not work.
yuou have to change the configuration value HashMode:
value 0: default hash crc if supportes fallback to mumurhash3
value 1: force murmurhash3
value 2: force xxhash
in my machine this are the results with al video settings to the minimun to evaluate cpu usage and using OGL:
CRC : 34-37 fps
murmurhash3 : 39-41 fps
xxhash: 40-42fps
to review the change see here
https://github.com/Tinob/dolphin/commit/741f4f271ccea397de2b9c5ca5d45d742b095f50

**Comment by degasus** [comment]
Sounds like xxhash is the way to go. It's BSD, so is it fine to inline into dolphin just as it is done in Tinob@741f4f2 or do I have to create a new external folder for it?

**Comment by BhaaLseN** [comment]
Hmm, thats interresting, we benchmarked xxhash a year (or so) back, and didn't really see any benefit over murmur.
Can you (or anyone else) also see those improvements with it?

**Comment by Tinob** [comment]
Forguet to say that my cpu is AMD fx-8350, it will be interesting to see results from other cpus. The best will be to test the bench on games with heavy texture usage.

**Comment by BhaaLseN** [comment]
Back in 2013, @lioncash implemented it as a replacement for murmur, and we concluded that it didn't make much of a difference. On my Phenom II X4 905e it was mostly either equal or xxhash slightly slower; with very few instances where it was faster (by a margin that could might aswell just be within the statistical error range of the measurements).

**Comment by Tinob** [comment]
Murmur is the winner then :)

**Comment by ZephyrSurfer** [comment]
On my Haswell G3258 I can only recommend using xxhash over Murmur  
I've done a lot of testing with this on different backends/ Audio or without/ different games/ dual core||single core.
You're right intel is good with CRC.  
This is one of the examples (D3D/ Single core/ No skipping/ No audio/ A fifolog)
0: CRC:                71 FPS
1: Murmurhash3: 62 FPS
2: xxhash:            68 FPS

**Comment by degasus** [comment]
So xxhash64 sounds like the fastest one. I'll have to wait for PR #1898

**Comment by degasus** [comment]
So I think I'm done, new xxhash is in use.

Last call for final review and testing

**Comment by BigheadSMZ** [comment]
Out of curiosity. Is it safe to use the latest build in this PR to start converting textures before it gets merged to master so we can get a jump start on it? (...as in, hashes are not going to change again)

**Comment by degasus** [comment]
@BigheadSMZ I don't know why the name should change again, so I think it's fine to start to redump.

**Comment by degasus** [comment]
@BigheadSMZ Still dumping?

**Comment by ZephyrSurfer** [comment]
A couple questions:
Does XXHash have less collisions than what was used forced to safe?  
Would dumping take longer with XXhash in externals over a smaller block of code inlined in Dolphin?

**Comment by BigheadSMZ** [comment]
I dumped and renamed for about 6 hours straight, and I must say I never get tired of seeing everything textured all the time! Everything seems to work perfectly, I probably have at least 80% of the textures converted, all of them would probably take a full playthrough.

I do have a few questions that all tie into the same answer. Can the two formats work together? Like if I nailed all the paletted textures, could I still include other textures that were previously dumped from fast? Or should I wait until I have everything renamed to release the pack? I'm patient enough to do a full playthrough if I must, but if I can take a shortcut to offer people something quick in the meantime then I'll go that route.

**Comment by degasus** [comment]
@ZephyrSurfer The generated code would be the same, so we should just care about the maintainance.

@BigheadSMZ Of course, both formats will live together for a while. But as long as both formats are used, we have to hash the texture in both ways. Maybe with a bit more overhead. I think the next stable release must support the old texture format, but I want to remove it quite after the next stable release. I want to merge those max/min calculations of paletted textures into videocommon to reduces redundant uploads. This maybe fixes the metroid prime stuttering :D

**Comment by BigheadSMZ** [comment]
That sounds perfect to me! Both me and frozenwings on the forum are starting a playthrough, between both of us we should hopefully hit everything along the way so this backwards compatibility will be awesome until we finish, and hopefully eliminate that massive 2.8GB pack we currently have as soon as possible. All of this is irrelevant to this PR, so in regards to it, there are no issues I could find and everything is working, including renaming mipmaps and deleting duplicates. I can't think of anything else, so from my perspective it looks good to me.

**Comment by Tinob** [comment]
I think there is a small error in the generation of the old hash, you always executes the xor betwen tex and tlut hash but in the old code the xor was only executed when the format needed it.

**Comment by degasus** [comment]
@Tinob This is a xor with zero, so it should be fine. Else no non-paletted texture could be loaded.

**Comment by Tinob** [comment]
:) sorry i was thinking in nxor

**Pull request merged**